org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java

```diff
@@ -200,6 +203,12 @@ public int instrumentAll(final InputStream input,
 
 	private int instrumentZip(final InputStream input,
 			final OutputStream output, final String name) throws IOException {
+		if (generators.getLast() instanceof
+				OfflineInstrumentationCompanionAccessGenerator) {
+			generators
+					.addLast(new
+							OfflineInstrumentationCompanionAccessGenerator());
```

Comment:
@bjkail I'm not sure that we were giving guarantee that `instrumentAll` is thread safe, or that we demanded implementation of `IExecutionDataAccessorGenerator` to be thread safe. I personally tend to consider everything as non thread safe unless stated otherwise.

One alternative to preserve thread safety here - is to create a new instance of `Instrumenter` inside instead of usage of `LinkedList`. But this will mean that `Instrumenter` should be final, thus again will be possible to say that we are breaking something. However again not sure that we were giving guarantee that clients can extend `Instrumenter`.

Another alternative is to introduce private methods `instrument*` with generator as parameter. But not sure if it's worth it, especially that `OfflineInstrumentationCompanionAccessGenerator` is not thread safe and requires call at the end of instrumentation.

In any case I want to believe that we have an ability to change API contracts when we have no other choice or when complexity of preservation is high, at least by doing increment of version up to `0.8.0`.